### PR TITLE
Unit test `compute_constants()`

### PR DESCRIPTION
**Pull request by Weibye**
# Objective

One more step towards #92

This PR is a bit of a mixed bag, so let me know if I should split it up. 

The objective of this PR is to set us up so that we can start testing the flexbox algorithm.

## Context

While trying to cover the `compute_constants()` in tests, I discovered the code needed a bit refactoring and massaging in order to actually be testable. 

Here I realized that the different `resolve()` methods can be abstracted to a "Resolve" trait, and following the convention of `MaybeMath`, we now have `MaybeResolve` and `ResolveOrDefault`. 

These traits encapsulate behaviour where we need to resolve from a potentially context-dependent size / dimension into a context-independent size / dimensions. In other words:

1. A size can be one (`Dimension`) or two dimensional (`Size<Dimension>` / `Rect<Dimension>`)
2. These sizes can be fully defined on their own:
     1. `Dimension::Points(n)` tells us that this is `n` long
3. Or sizes can be dependent on the size of its parent
     1. `Dimension::Percent(n)` tells us that this is `n * parent.width` long
     2. _Hence "context-dependent"_
4. In order perform calculation of the node, we need the sizes to be fully independent of any parent context. This is where the `resolve` come in.

The only real difference between the two is what they output when they are unable to resolve the size
- `MaybeResolve` will return a `Option<f32>` with a `None` as the fallback
- `ResolveOrDefault` will return a `f32` with the value `0.0` as the fallback

## Changelog

- Created a new module `resolve` which houses all code related code
- Created the trait `MaybeResolve`
- Create the trait `ResolveOrDefault`
- `Dimension.resolve()` and `Size<Dimension>.resolve()` has been replaced by `impl MaybeResolve`
- Changed the signature of `compute_constants()` to make it easier to test
- Added `#[must_use]` on static constructors that was missing them
- Added convenience constructors on `Size<Dimension>`, `Rect<Dimension>`
- **[BREAKING]** `Size<f32>.zero()` has now become `Size::<f32>::ZERO`
- **[BREAKING]** `Point<f32>.zero()` has now become `Point::<f32>::ZERO`
- Added a `new()` constructor to `Size<f32>`
- Removed `Rect<T>.map` as not used anymore

## Feedback wanted

- Does the traits make sense and make the code cleaner + easier to understand?
- Docs help plz

### CODE REVIEW

File: src/geometry.rs [hunk 1]
```diff
@@ -231,16 +231,44 @@ impl Size<f32> {
...
+impl Size<Option<f32>> {
+    /// TODO
+    #[must_use]
+    pub fn from_none() -> Self {
+        Size { width: None, height: None }
+    }
+
+    /// TODO
+    #[must_use]
+    pub fn from_some(width: f32, height: f32) -> Self {
```
Comment (alice-i-cecile) [inline src/geometry.rs:243]: I definitely think the name could be improved here. Not immediately sure how.

File: src/geometry.rs [hunk 2]
```diff
@@ -231,16 +231,44 @@ impl Size<f32> {
     }
 }
 
+impl Size<Option<f32>> {
+    /// TODO
+    #[must_use]
+    pub fn from_none() -> Self {
```
Comment (alice-i-cecile) [inline src/geometry.rs:237]: This should be an associated constant, not a method.

File: src/geometry.rs (hunk 1 above)
Comment (alice-i-cecile) [reply src/geometry.rs:243]: Hmm. This feels very similar to the `from_points` method below?

Comment (Weibye) [reply src/geometry.rs:243]: Yeah, not fully sure what is a better naming. This constructs a `Size<Option<f32>>`, `from_points()` constructs a `Size<Dimension::Points(f32)>`

Perhaps the biggest downside is what to do when you need to create a method that create a `Size<Option<OtherType>>` in which case `from_some()` would be wholly ambiguous from naming alone.

Comment (alice-i-cecile) [reply src/geometry.rs:243]: We could just call it `new`? This feels like the default constructor.

**Comment by Weibye** [comment]
Updated the PR description to reflect the work done so far

File: RELEASES.md [hunk 3]
```diff
@@ -9,6 +9,8 @@
 
 ### 0.2.0 Changed
 
+- **[BREAKING]** `Size<f32>.zero()` has now become `Size::<f32>::ZERO`
```
Comment (alice-i-cecile) [inline RELEASES.md:12]:
```suggestion
- **[BREAKING]** `Size<f32>.zero()` is now `Size::<f32>::ZERO`
```

File: RELEASES.md [hunk 4]
```diff
@@ -9,6 +9,8 @@
...
+- **[BREAKING]** `Point<f32>.zero()` has now become `Point::<f32>::ZERO`
```
Comment (alice-i-cecile) [inline RELEASES.md:13]:
```suggestion
- **[BREAKING]** `Point<f32>.zero()` is now  `Point::<f32>::ZERO`
```

File: src/geometry.rs [hunk 5]
```diff
@@ -32,6 +32,7 @@ impl<T> Rect<T> {
     /// Applies the function `f` to all four sides of the [`Rect`]
     ///
     /// This is used to transform a `Rect<T>` into a `Rect<R>`.
+    #[allow(dead_code)]
```
Comment (alice-i-cecile) [inline src/geometry.rs:35]: We should just get rid of this code instead.

File: src/resolve.rs [hunk 6]
```diff
@@ -293,4 +303,62 @@ mod tests {
     }
 
     // TODO: Add tests for ResolveOrDefault<Option<f32>, Rect<f32>> for Rect<Dimension>
+    mod resolve_or_default_rect_dimension_to_rect_f32_via_option {
+        use crate::geometry::{Rect, Size};
```
Comment (alice-i-cecile) [inline src/resolve.rs:307]:
```suggestion
        use crate::geometry::Rect;
```

File: RELEASES.md [hunk 7]
```diff
@@ -9,6 +9,8 @@
 
 ### 0.2.0 Changed
 
+- **[BREAKING]** `Size<f32>.zero()` is now `Size::<f32>::ZERO`
```
Comment (alice-i-cecile) [inline RELEASES.md:12]:
```suggestion
- `Size<f32>.zero()` is now `Size::<f32>::ZERO`
```

Comment (alice-i-cecile) [reply RELEASES.md:12]: These are breaking by default :)

File: RELEASES.md [hunk 8]
```diff
@@ -9,6 +9,8 @@
...
+- **[BREAKING]** `Point<f32>.zero()` is now  `Point::<f32>::ZERO`
```
Comment (alice-i-cecile) [inline RELEASES.md:13]:
```suggestion
- `Point<f32>.zero()` is now  `Point::<f32>::ZERO`
```

**Review by alice-i-cecile — APPROVED** [review]
Looks good to me now. Are you happy with this?

**Comment by Weibye** [comment]
> Looks good to me now. Are you happy with this?

Yes! I'm very happy with how this turned out, and I think this is the right place to stop ^^

**Pull request merged**